### PR TITLE
consensus: p2p refactor

### PR DESCRIPTION
**Pull request by alexanderbez**
## Description

Prepares the consensus reactor for the newly designed p2p changes per [ADR 062](https://github.com/tendermint/tendermint/blob/master/docs/architecture/adr-062-p2p-architecture.md).

- Move `PeerState` to `peer_state.go` to improve readability and separate concerns in `reactor.go`.
  - Add `running`, `closer`, and `broadcastWG` fields to synchronize spawned goroutines when adding/removing a peer.
- Replicate `p2p.Peer#Get|Set` by making `Reactor` have an internal `peers *cmap.CMap` where we now store `PeerState`.
- Update a few p2p APIs to allow the use of buffered channels.

**Note to reviewers**: To simplify review, I would mainly focus on changes to `reactor.go` and `peer_state.go` and perhaps some of the messages changes.

ref: #5670

### CODE REVIEW

**Comment by alexanderbez** [comment]
Ok, so the reactor refactor itself is pretty much done. I don't know how long the tests will take me to refactor -- they seem very...messy. But I'm hopeful I can get most of them done by this week.

In the interest of time, feel free to just review the core changes (i.e. `reactor.go` and the auxiliary/non-test-related files ) @melekes @erikgrinaker.

**Comment by erikgrinaker** [comment]
Great stuff, will have a look later today!

**Comment by alexanderbez** [comment]
Thanks @erikgrinaker. I think this is pretty standard stuff as far as reviewal compared to the other refactors (i.e. it's pretty much the same).

What I would probably spend the most effort on reviewing is the usage of `PeerState` and how we manage the goroutines there (i.e. the waitgroup).

File: consensus/reactor.go [hunk 1]
```diff
@@ -557,1170 +536,797 @@ OUTER_LOOP:
...
-// NOTE: For now, it is just a hard-coded string to avoid accessing unprotected shared variables.
-// TODO: improve!
-func (conR *Reactor) String() string {
-	// better not to access shared variables
-	return "ConsensusReactor" // conR.StringIndented("")
-}
+		// save the peer if the ID is non-empty and we haven't seen this peer before
+		if psKey != "" && !r.peers.Has(psKey) {
+			peerState := NewPeerState(r.Logger, peerUpdate.PeerID)
+			r.peers.Set(psKey, peerState)
```
Comment (erikgrinaker) [inline consensus/reactor.go:857]: There's a race condition here, since `Has()`, `Get()`, and `Set()` are not called in the same critical section (i.e. while holding a mutex). It's possible for a different goroutine to call `Delete()` on the peer between `Has()` and `Get()`, causing the reactor to panic. There's a few different ways to handle this, depending on our needs, but in this case it's probably sufficient to add e.g. `peers.Ensure()` which atomically returns the existing peer state if any or otherwise creates a new one, stores it in the map, and returns it.

File: consensus/peer_state.go [hunk 2]
```diff
@@ -0,0 +1,522 @@
...
+	ps.running = v
+}
+
+// IsRunning returns true if a PeerState is considered running where multiple
+// broadcasting goroutines exist for the peer.
+func (ps *PeerState) IsRunning() bool {
+	ps.mtx.RLock()
+	defer ps.mtx.RUnlock()
+
+	return ps.running
```
Comment (erikgrinaker) [inline consensus/peer_state.go:83]: I'm a bit worried about having two different signals for shutdown here. Can we just rely on `ps.closer` for `IsRunning()` too?

File: consensus/reactor.go [hunk 3]
```diff
@@ -760,967 +656,688 @@ func (conR *Reactor) gossipVotesForHeight(
...
+			// longer running.
+			//
+			// XXX/TODO: It is possible that we can briefly wait here for all goroutines.
+			// Is this problematic?
+			peerState.closer.Close()
+			peerState.broadcastWG.Wait()
+			peerState.SetRunning(false)
+
+			// remove peer from the peer store to prevent unbounded growth
+			r.peers.Delete(psKey)
```
Comment (erikgrinaker) [inline consensus/reactor.go:916]: Maybe we should remove from the map before signalling shutdown, to prevent anything else from touching the peer state while it's shutting down?

File: consensus/reactor.go [hunk 4]
```diff
@@ -760,967 +656,688 @@ func (conR *Reactor) gossipVotesForHeight(
...
-	ps.mtx.Lock()
-	defer ps.mtx.Unlock()
-	return ps.PRS.Height
-}
+// handleStateMessage handles envelopes sent from peers on the StateChannel.
+// An error is returned if the message is unrecognized or if validation fails.
+func (r *Reactor) handleStateMessage(envelope p2p.Envelope, msgI Message) error {
+	ps, ok := r.peers.Get(string(envelope.From)).(*PeerState)
+	if !ok {
+		panic(fmt.Sprintf("peer %v has no state", envelope.From))
```
Comment (erikgrinaker) [inline consensus/reactor.go:926]: In this case I think we should just return rather than panicing. If the channel is backed up it's quite possible to receive a message from a peer after the PeerStatusDown message is processed, since these channels are not ordered with respect to each other. This goes for other locations as well.

**Review by erikgrinaker — COMMENTED** [review]
Great that the consensus reactor is getting some ❤️, it's quite a bit easier to navigate now that it's split up like this.

File: consensus/reactor.go (hunk 1 above)
Comment (alexanderbez) [reply consensus/reactor.go:874]: `r.peers` is thread-safe, no? So why do we need an additional mutex?

File: consensus/reactor.go (hunk 4 above)
Comment (alexanderbez) [reply consensus/reactor.go:926]: Yeah, good catch.

File: consensus/reactor.go (hunk 1 above)
Comment (alexanderbez) [reply consensus/reactor.go:874]: Ohh you mean a mutex around the entire block of execution -- not just API calls to `peers`?

Comment (erikgrinaker) [reply consensus/reactor.go:874]: No wait, this actually isn't true, since `processPeerUpdate` is not called concurrently and it is the only method that mutates the map.

Either way, I'd simplify the code a bit by doing `peerState, ok := Get(); if !ok { peerState = NewPeerState(); Set() }`, which avoids a panic case.

File: consensus/reactor.go [hunk 5]
```diff
@@ -760,967 +662,688 @@ func (conR *Reactor) gossipVotesForHeight(
...
+		peerState, ok := r.peers.Get(psKey).(*PeerState)
+		if ok && peerState.IsRunning() {
+			// Close the closer to signal to all spawned goroutines to gracefully exit
+			// and we wait for them to all exit before marking the peer state as no
+			// longer running.
+			//
+			// XXX/TODO: It is possible that we can briefly wait here for all goroutines.
+			// Is this problematic?
+			peerState.closer.Close()
+			peerState.broadcastWG.Wait()
```
Comment (erikgrinaker) [inline consensus/reactor.go:912]: This can wait for a long time, which will block `processPeerUpdate()` from processing further updates, potentially stalling the `PeerManager`. Might be better to just fetch the state and remove it from the map, then spawn a goroutine to stop the peerState.

Comment (alexanderbez) [reply consensus/reactor.go:912]: Yeah I've gone back-n-forth on this. This is the only bit of code I don't like. I'll marinate on your suggestion a bit 👍

File: consensus/peer_state.go (hunk 2 above)
Comment (alexanderbez) [reply consensus/peer_state.go:83]: How would that work? Use `ps.closer.Done()`? Wouldn't that block?

I don't think the current method is too bad.

Comment (erikgrinaker) [reply consensus/peer_state.go:83]: You'd do a non-blocking read to check whether it was closed or not.

Either way, since this isn't concurrent it's not that important -- but if it was, there's be issues due to non-atomicity so you'd have to be super-careful about the ordering of operations and checks.

File: consensus/reactor.go [hunk 6]
```diff
@@ -1,550 +1,535 @@
...
+		r.conS.Wait()
 	}
-	if !conR.WaitSync() {
-		conR.conS.Wait()
+
+	// wait for all spawned peer goroutines to gracefully exit
+	r.mtx.Lock()
+	for _, ps := range r.peers {
+		ps.closer.Close()
+		ps.broadcastWG.Wait()
```
Comment (erikgrinaker) [inline consensus/reactor.go:198]: I don't think we should wait while holding the mutex. Maybe just grab a copy of the map while holding the mutex, then iterate once to close, then iterate again to wait (so they close concurrently).

File: consensus/reactor.go [hunk 7]
```diff
@@ -760,967 +662,720 @@ func (conR *Reactor) gossipVotesForHeight(
...
+
+			// start goroutines for this peer
+			go r.gossipDataRoutine(ps)
+			go r.gossipVotesRoutine(ps)
+			go r.queryMaj23Routine(ps)
+
+			// Send our state to the peer. If we're fast-syncing, broadcast a
+			// RoundStepMessage later upon SwitchToConsensus().
+			if !r.waitSync {
+				r.sendNewRoundStepMessage(ps.peerID)
```
Comment (erikgrinaker) [inline consensus/reactor.go:904]: We may want to consider running this as a goroutine, since it grabs the conS.mtx lock and thus may block for a while as it's waiting for it.

File: consensus/reactor.go [hunk 8]
```diff
@@ -760,967 +662,720 @@ func (conR *Reactor) gossipVotesForHeight(
...
+		ps, ok := r.peers[peerUpdate.PeerID]
+		if ok && ps.IsRunning() {
+			// signal to all spawned goroutines for the peer to gracefully exit
+			ps.closer.Close()
+
+			go func() {
+				// Wait for all spawned broadcast goroutines to exit before marking the
+				// peer state as no longer running and removal from the peers map.
+				ps.broadcastWG.Wait()
+				ps.SetRunning(false)
```
Comment (erikgrinaker) [inline consensus/reactor.go:918]: We should delete from the map before setting running to false, otherwise there is a race where e.g. we set running=false, the main thread picks the peer state from the map, sees that it is not running, and starts it again.

File: consensus/reactor.go (hunk 7 above)
Comment (alexanderbez) [reply consensus/reactor.go:904]: Yes, makes sense since we send on a p2p channel. Good catch!

File: consensus/msgs.go [hunk 9]
```diff
@@ -4,18 +4,319 @@ import (
...
+	if m.LastCommitRound < -1 {
+		return errors.New("invalid LastCommitRound (cannot be < -1)")
+	}
+
+	return nil
+}
+
+// ValidateHeight validates the height given the chain's initial height.
+func (m *NewRoundStepMessage) ValidateHeight(initialHeight int64) error {
+	if m.Height < initialHeight {
```
Comment (melekes) [inline consensus/msgs.go:72]: maybe use a `switch` here?

File: consensus/peer_state.go [hunk 10]
```diff
@@ -0,0 +1,522 @@
...
+	"github.com/tendermint/tendermint/libs/log"
+	tmsync "github.com/tendermint/tendermint/libs/sync"
+	"github.com/tendermint/tendermint/p2p"
+	tmproto "github.com/tendermint/tendermint/proto/tendermint/types"
+	"github.com/tendermint/tendermint/types"
+	tmtime "github.com/tendermint/tendermint/types/time"
+)
+
+var (
+	ErrPeerStateHeightRegression = errors.New("error peer state height regression")
```
Comment (melekes) [inline consensus/peer_state.go:21]:
```suggestion
	ErrPeerStateHeightRegression = errors.New("peer state height regression")
```

File: consensus/peer_state.go [hunk 11]
```diff
@@ -0,0 +1,522 @@
...
+	ErrPeerStateInvalidStartTime = errors.New("error peer state invalid startTime")
```
Comment (melekes) [inline consensus/peer_state.go:22]:
```suggestion
	ErrPeerStateInvalidStartTime = errors.New("peer state invalid startTime")
```

File: consensus/peer_state.go [hunk 12]
```diff
@@ -0,0 +1,522 @@
...
+// IsRunning returns true if a PeerState is considered running where multiple
+// broadcasting goroutines exist for the peer.
+func (ps *PeerState) IsRunning() bool {
+	ps.mtx.RLock()
+	defer ps.mtx.RUnlock()
+
+	return ps.running
+}
+
+// GetRoundState returns an shallow copy of the PeerRoundState. There's no point
```
Comment (melekes) [inline consensus/peer_state.go:86]:
```suggestion
// GetRoundState returns a shallow copy of the PeerRoundState. There's no point
```

File: consensus/peer_state.go [hunk 13]
```diff
@@ -0,0 +1,522 @@
...
+	}
+
+	ps.PRS.ProposalBlockParts.SetIndex(index, true)
+}
+
+// PickVoteToSend picks a vote to send to the peer. It will return true if a
+// vote was picked.
+//
+// NOTE: `votes` must be the correct Size() for the Height().
+func (ps *PeerState) PickVoteToSend(votes types.VoteSetReader) (vote *types.Vote, ok bool) {
```
Comment (melekes) [inline consensus/peer_state.go:169]: not sure why we need `ok bool` here

File: consensus/peer_state.go [hunk 14]
```diff
@@ -0,0 +1,522 @@
...
+	ps.mtx.Lock()
+	defer ps.mtx.Unlock()
+
+	// ignore duplicates or decreases
+	if CompareHRS(msg.Height, msg.Round, msg.Step, ps.PRS.Height, ps.PRS.Round, ps.PRS.Step) <= 0 {
+		return
+	}
+
+	// Just remember these values.
+	psHeight := ps.PRS.Height
```
Comment (melekes) [inline consensus/peer_state.go:394]:
```
var (
)
```

File: consensus/reactor.go [hunk 15]
```diff
@@ -1,550 +1,533 @@
...
+func (r *Reactor) StringIndented(indent string) string {
+	s := "ConsensusReactor{\n"
+	s += indent + "  " + r.conS.StringIndented(indent+"  ") + "\n"
 
-// AddPeer implements Reactor by spawning multiple gossiping goroutines for the
-// peer.
-func (conR *Reactor) AddPeer(peer p2p.Peer) {
-	if !conR.IsRunning() {
-		return
+	for _, ps := range r.peers {
```
Comment (melekes) [inline consensus/reactor.go:286]: do we need to grab a read lock here?

**Review by melekes — COMMENTED** [review]
looking good so far 👍

File: consensus/reactor_test.go [hunk 16]
```diff
@@ -3,623 +3,822 @@ package consensus
...
-	reactors, blocksSubs, eventBuses := startConsensusNet(t, css, nPeers)
-	defer stopConsensusNet(logger, reactors, eventBuses)
-
-	// map of active validators
-	activeVals := make(map[string]struct{})
-	for i := 0; i < nVals; i++ {
-		pubKey, err := css[i].privValidator.GetPubKey()
-		require.NoError(t, err)
-		activeVals[string(pubKey.Address())] = struct{}{}
+func simulateRouter(primary *reactorTestSuite, suites []*reactorTestSuite, dropChErr bool) {
```
Comment (erikgrinaker) [inline consensus/reactor_test.go:142]: I think this should use an actual router with in-memory transports instead, since these are basically integration tests. I'll be writing some more test infrastructure for setting up these kinds of networks shortly.

Comment (alexanderbez) [reply consensus/reactor_test.go:142]: Perhaps. I have all reactors using this kind of approach. Once we have an in memory router setup that I can leverage, I'd be happy to switch it over to that 👍

Comment (erikgrinaker) [reply consensus/reactor_test.go:142]: As discussed elsewhere, I think we should primarily rely on unit tests rather than integration tests for reactors, since they're much more specific, can cover more ground with less effort, and are less prone to race conditions. But we'll need some of both, obviously.

Comment (alexanderbez) [reply consensus/reactor_test.go:142]: Do we have an in-mem transport/router that I can already leverage? If so, can you point me to its use?

Comment (erikgrinaker) [reply consensus/reactor_test.go:142]: There's some stuff in `p2p/router_test.go`, but we need some better convenience test infra for it, and I'm currently overhauling the transports a bit anyway. Will have them ready over the weekend.

File: consensus/peer_state.go (hunk 13 above)
Comment (alexanderbez) [reply consensus/peer_state.go:169]: remnants from the old code

File: consensus/msgs.go (hunk 9 above)
Comment (alexanderbez) [reply consensus/msgs.go:72]: `switch` would be useful for else/if blocks.

File: consensus/reactor.go (hunk 15 above)
Comment (alexanderbez) [reply consensus/reactor.go:290]: Yeah :)

File: consensus/reactor_test.go (hunk 16 above)
Comment (erikgrinaker) [reply consensus/reactor_test.go:151]: I've added some test tooling in `p2p/p2ptest`, including setting up in-memory networks for testing and a bunch of channel-based test assertions. This has mostly been geared towards router testing, and needs to be extended and/or adapted to reactor-based testing, but it's a start. You can see them used e.g. in `p2p/router_test.go`.

File: p2p/queue.go [hunk 17]
```diff
@@ -33,9 +33,9 @@ type fifoQueue struct {
 
 var _ queue = (*fifoQueue)(nil)
 
-func newFIFOQueue() *fifoQueue {
+func newFIFOQueue(size int) *fifoQueue {
```
Comment (alexanderbez) [inline p2p/queue.go:36]: /cc @erikgrinaker -- in order to get this to play nicely with the p2p test network stack and the consensus reactor, we need these channels to be buffered. In the router, of course, we maintain them to be unbuffered (i.e. 0).

Not sure if you had thoughts on this as far as APIs and what should be buffered, how much, and where.

File: p2p/router.go [hunk 18]
```diff
@@ -238,11 +238,12 @@ func NewRouter(
...
-func (r *Router) OpenChannel(id ChannelID, messageType proto.Message) (*Channel, error) {
-	queue := newFIFOQueue()
-	outCh := make(chan Envelope)
-	errCh := make(chan PeerError)
+// wrapper message. The caller may provide a size to make the channel buffered,
+// which internally makes the inbound, outbound, and error channel buffered.
+func (r *Router) OpenChannel(id ChannelID, messageType proto.Message, size int) (*Channel, error) {
+	queue := newFIFOQueue(size)
+	outCh := make(chan Envelope, size)
+	errCh := make(chan PeerError, size)
```
Comment (alexanderbez) [inline p2p/router.go:246]: /cc @erikgrinaker -- in order to get this to play nicely with the p2p test network stack and the consensus reactor, we need these channels to be buffered. In the router, of course, we maintain them to be unbuffered (i.e. 0).

Not sure if you had thoughts on this as far as APIs and what should be buffered, how much, and where.

File: p2p/p2ptest/network.go [hunk 19]
```diff
@@ -34,16 +34,24 @@ func MakeNetwork(t *testing.T, nodes int) *Network {
...
 		network.Nodes[node.NodeID] = node
 	}
 
+	return network
+}
+
+// Start starts the network by setting up a list of node addresses to dial in
+// addition to creating a peer update subscription for each node. Finally, all
+// nodes are connected to each other.
+func (n *Network) Start(t *testing.T) {
```
Comment (alexanderbez) [inline p2p/p2ptest/network.go:49]: note to reviewer: `Start` was added because in reactor tests, we want to fully _create and start_ the reactor before getting full peer connectivity.

File: p2p/router.go (hunk 18 above)
Comment (erikgrinaker) [reply p2p/router.go:246]: Sure. You could consider buffering the `MemoryConnection` queues instead, since all networked transports (including `MConnTransport`) does a bunch of internal buffering already, as does the OS networking stack, so this gives more consistent behavior across transports.

That said, any non-trivial `queue` implementations will need internal buffers too (because they need a set of messages to apply QoS algorithms to).

Comment (alexanderbez) [reply p2p/router.go:246]: > That said, any non-trivial queue implementations will need internal buffers too (because they need a set of messages to apply QoS algorithms to).

This encourages me to leave this as-is for now. Although I do agree we should allow the `MemoryConnection` to be configurable.

File: consensus/reactor.go [hunk 20]
```diff
@@ -1,551 +1,561 @@
...
+	stateCh       *p2p.Channel
+	dataCh        *p2p.Channel
+	voteCh        *p2p.Channel
+	voteSetBitsCh *p2p.Channel
+	peerUpdates   *p2p.PeerUpdates
+
+	// NOTE: We need a dedicated stateCloseCh channel for signaling closure of
+	// the StateChannel due to the fact that the StateChannel message handler
+	// performs a send on the VoteSetBitsChannel. This is an antipattern, so having
+	// this dedicated channel,stateCloseCh, is necessary in order to avoid data races.
```
[the remark this answers is not in the document]
Comment (cmwaters) [reply consensus/reactor.go:114]: Also just so I understand correctly, we need the separate channel because we need to make sure we close state channel before the rest of the channels right?

File: consensus/reactor.go [hunk 21]
```diff
@@ -1,551 +1,561 @@
...
+	StateChannel       = p2p.ChannelID(0x20)
+	DataChannel        = p2p.ChannelID(0x21)
+	VoteChannel        = p2p.ChannelID(0x22)
+	VoteSetBitsChannel = p2p.ChannelID(0x23)
 
-	maxMsgSize = 1048576 // 1MB; NOTE/TODO: keep in sync with types.PartSet sizes.
+	maxMsgSize = 1048576 // 1MB; NOTE: keep in sync with types.PartSet sizes.
 
 	blocksToContributeToBecomeGoodPeer = 10000
 	votesToContributeToBecomeGoodPeer  = 10000
```
Comment (cmwaters) [inline consensus/reactor.go:86]: What's the role of something like this with respect to the new framework

**Review by cmwaters — APPROVED** [review]
utACK. Really good work! 🚀  (my comments are all questions)

In terms of the design, I couldn't think of any scenarios where we might leak go routines. The ordering of the execution of processes seems sound. I also like the extraction of peer state into its own file. I guess the only qualm I have is that we have two stats that we use to keep track of peer behavior (votes and blocks) but the peer manager is never informed when anything bad happens i.e. invalid proposal or invalid vote.

File: consensus/reactor.go (hunk 21 above)
Comment (alexanderbez) [reply consensus/reactor.go:86]: Good question. So in `peerStatsRoutine` we use these variables to invoke the (legacy) `MarkPeerAsGood` API on the switch, which internally does some logic within the address book.

In the new framework, the `PeerManager` has a notion of a "score" for each peer. These scores are currently _binary_ (0 or 100). While it's not clear how exactly to implement a more sophisticated scoring system, I think the idea will be to either still have these APIs or piggy-back off of the `PeerUpdate` mechanism to send scoring updates.

tl;dr `MarkPeerAsGood` -> Update the peer score. How the score is updated remains to be determined.

File: consensus/reactor.go (hunk 20 above)
[the remark this answers is not in the document]
Comment (alexanderbez) [reply consensus/reactor.go:114]: > What's the reason behind the state channel sending stuff also over the vote set bits channel? Is it a timing thing? Like only on certain conditions do we sent vote set bits?

Good question. To me, it seems like a code smell. i.e. you should only send and receive messages on dedicated channels and not mix-n-match. As to the why, unfortunately idk. This is an artifact of how this reactor was designed/implemented. Perhaps @ebuchman can elaborate on why we do this?

**Comment by alexanderbez** [comment]
>  I guess the only qualm I have is that we have two stats that we use to keep track of peer behavior (votes and blocks) but the peer manager is never informed when anything bad happens i.e. invalid proposal or invalid vote.

Indeed this is true, but I'm not sure where/if this happens. I did a quick skim of the code and I couldn't see where this happens in the reactor or the state. Do you happen to know?

**Comment by cmwaters** [comment]
Yeah look at this function here
```
func (cs *State) handleMsg(mi msgInfo) {
```

In it we have `setProposal()` and `tryAddVote()` in which the error messages are just logged but there should be some peer handling done here

**Comment by alexanderbez** [comment]
Ahh yes. Perhaps we can piggy-back off of `msgInfo` and add an `Error` field, when non-nil, `peerStatsRoutine` can send an update to penalize the peer.

File: consensus/reactor.go [hunk 22]
```diff
@@ -762,967 +688,701 @@ func (conR *Reactor) gossipVotesForHeight(
...
+		if ok && ps.IsRunning() {
+			// signal to all spawned goroutines for the peer to gracefully exit
+			ps.closer.Close()
 
-	ps.PRS.ProposalBlockPartSetHeader = proposal.BlockID.PartSetHeader
-	ps.PRS.ProposalBlockParts = bits.NewBitArray(int(proposal.BlockID.PartSetHeader.Total))
-	ps.PRS.ProposalPOLRound = proposal.POLRound
-	ps.PRS.ProposalPOL = nil // Nil until ProposalPOLMessage received.
-}
+			go func() {
```
Comment (melekes) [inline consensus/reactor.go:945]: because we do it in a goroutine, is it possible to get another PeerStatusDown for the same peer and end up blocking in a second goroutine on `ps.broadcastWG.Wait()`, therefore leaking 1 goroutine?

File: consensus/reactor.go [hunk 23]
```diff
@@ -762,967 +688,701 @@ func (conR *Reactor) gossipVotesForHeight(
...
 
-// String returns a string representation.
-func (m *VoteSetMaj23Message) String() string {
-	return fmt.Sprintf("[VSM23 %v/%02d/%v %v]", m.Height, m.Round, m.Type, m.BlockID)
-}
+func (r *Reactor) peerStatsRoutine() {
+	for {
+		if !r.IsRunning() {
+			r.Logger.Info("stopping peerStatsRoutine")
+			return
```
Comment (melekes) [inline consensus/reactor.go:1359]: should we flush `statsMsgQueue` before exiting?

**Review by melekes — COMMENTED** [review]
nice work 👍 left a few questions, otherwise LGTM

**Comment by cmwaters** [comment]
> Ahh yes. Perhaps we can piggy-back off of msgInfo and add an Error field, when non-nil, peerStatsRoutine can send an update to penalize the peer.

This would seem like the most logical place to return the error. I don't understand why we return the entire message back to the reactor when we're just recording votes and block parts. I would almost be tempted to just have a new struct for reporting peer stats/behavior. 

The other thing that bothers me is this:
```
// if err == ErrAddingVote {
	// TODO: punish peer
	// We probably don't want to stop the peer here. The vote does not
	// necessarily comes from a malicious peer but can be just broadcasted by
	// a typical peer.
	// https://github.com/tendermint/tendermint/issues/1281
// }
```
Why do we continue to broadcast invalid votes? 
I'll let you decide if you want to implement anything now or just make a note of this somewhere and we can address it at a later point

File: consensus/peer_state.go [hunk 24]
```diff
@@ -0,0 +1,528 @@
+package consensus
```
[the remark this answers is not in the document]
Comment (alexanderbez) [reply consensus/peer_state.go:1]: Yes, it was moved from `reactor.go` to its own dedicated file `peer_state.go`, which I believe makes things easier to read and allows for separation of concerns. Note, there are a few actual tweaks to `PeerState` outlined in the PR body.

**Comment by alexanderbez** [comment]
@cmwaters can you comment on the issue I've created? You bring up good points and I don't want them to get buried in this PR (as to also not bike-shed).

File: consensus/reactor.go (hunk 22 above)
Comment (alexanderbez) [reply consensus/reactor.go:945]: > is it possible to get another PeerStatusDown for the same peer

Yes. But I don't think this is a concern, one or both or many goroutines will block on `ps.broadcastWG.Wait()`, but once that waitgroup hits zero, they should all complete and the remaining logic becomes idempotent.

File: consensus/reactor.go (hunk 23 above)
Comment (alexanderbez) [reply consensus/reactor.go:1359]: We can execute `close(statsMsgQueue)` during `State#OnStop`, so it's closed when the reactor closes. WDYT?

Comment (alexanderbez) [reply consensus/reactor.go:1359]: Then again, the `State` type doesn't close any of its channels. Not sure if its worth it?

Comment (melekes) [reply consensus/reactor.go:1359]: I don't think we should close statsMsgQueue - will keep us away from panicking. You mean old implementation does not do flushing? Then I guess it's fine. But this is what I would've done to ensure all goroutines that are writing to statsMsgQueue terminate.

Comment (alexanderbez) [reply consensus/reactor.go:1359]: > You mean old implementation does not do flushing? 

Yes, correct. In fact, `peerStatsRoutine` was largely left untouched.
